File: packages/react-native-audio-api/android/src/main/java/com/swmansion/audioapi/nodes/StereoPannerNode.kt
package com.swmansion.audioapi.nodes

import android.os.Build
import androidx.annotation.RequiresApi
import com.swmansion.audioapi.context.BaseAudioContext
import com.swmansion.audioapi.nodes.audionode.AudioNode
import com.swmansion.audioapi.nodes.audionode.ChannelCountMode
import com.swmansion.audioapi.parameters.AudioParam
import com.swmansion.audioapi.parameters.PlaybackParameters
import com.swmansion.audioapi.utils.Constants
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.sin

// https://webaudio.github.io/web-audio-api/#stereopanner-algorithm

class StereoPannerNode(
  context: BaseAudioContext,
) : AudioNode(context) {
  override val numberOfInputs: Int = 1
  override val numberOfOutputs: Int = 1
  override val channelCountMode: ChannelCountMode = ChannelCountMode.CLAMPED_MAX

  private val pan: AudioParam = AudioParam(context, 0.0, Constants.MAX_PAN, -Constants.MAX_PAN)
    get() = field

  @RequiresApi(Build.VERSION_CODES.N)
  override fun process(playbackParameters: PlaybackParameters) {
    mixBuffers(playbackParameters)

    val bufferL = playbackParameters.audioBuffer.getChannelData(0)
    val bufferR = playbackParameters.audioBuffer.getChannelData(1)

    for (i in 0 until playbackParameters.audioBuffer.length) {
      val pan = pan.getValueAtTime(context.getCurrentTime())
      val x = (if (pan <= 0) pan + 1 else pan) * PI / 2

      val gainL = cos(x)
      val gainR = sin(x)

      val inputL = bufferL[i]
      val inputR = bufferR[i]

      if (pan <= 0) {
        bufferL[i] += (inputR * gainL).toFloat()
        bufferR[i] *= (gainR).toFloat()
      } else {
        bufferL[i] *= (gainL).toFloat()
        bufferR[i] += (inputL * gainR).toFloat()
      }
    }

    super.process(playbackParameters)
  }
}

File: packages/react-native-audio-api/ios/nodes/StereoPannerNode/StereoPannerNode.m
#import "StereoPannerNode.h"
#import "AudioContext.h"

// https://webaudio.github.io/web-audio-api/#stereopanner-algorithm

@implementation StereoPannerNode

- (instancetype)initWithContext:(AudioContext *)context
{
  if (self = [super initWithContext:context]) {
    _panParam = [[AudioParam alloc] initWithContext:context value:0 minValue:-1 maxValue:1];
    self.numberOfInputs = 1;
    self.numberOfOutputs = 1;
    self.channelCountMode = ChannelCountModeClampedMax;
  }

  return self;
}

- (void)cleanup
{
  _panParam = nil;
}

- (void)process:(AVAudioFrameCount)frameCount bufferList:(AudioBufferList *)bufferList
{
  float time = [self.context getCurrentTime];
  float deltaTime = 1 / self.context.sampleRate;

  float *bufferL = (float *)bufferList->mBuffers[0].mData;
  float *bufferR = (float *)bufferList->mBuffers[1].mData;

  for (int frame = 0; frame < frameCount; frame += 1) {
    float pan = [_panParam getValueAtTime:[self.context getCurrentTime]];
    float x = ((pan <= 0) ? pan + 1 : pan) * M_PI / 2;

    float gainL = cos(x);
    float gainR = sin(x);

    float inputL = bufferL[frame];
    float inputR = bufferR[frame];

    if (pan <= 0) {
      bufferL[frame] += inputR * gainL;
      bufferR[frame] *= gainR;
    } else {
      bufferL[frame] *= gainL;
      bufferR[frame] += inputL * gainR;
    }

    time += deltaTime;
  }

  [super process:frameCount bufferList:bufferList];
};

@end
